### PR TITLE
improvement(k8s): split K8S node diruptive nemesis to separate methods

### PR DESCRIPTION
**Pull request by vponomaryov**
Changes:
- Split `disrupt_terminate_and_replace_node_kubernetes` into following:
  - `disrupt_drain_kubernetes_node_then_replace_scylla_node`
  - `disrupt_terminate_kubernetes_host_then_replace_scylla_node`
- Split `disrupt_terminate_decommission_add_node_kubernetes` into following:
  - `disrupt_drain_kubernetes_node_then_decommission_and_add_scylla_node`
  - `disrupt_terminate_kubernetes_host_then_decommission_and_add_scylla_node`
- Rename `disrupt_replace_node_kubernetes` method -> `disrupt_replace_scylla_node_on_kubernetes`.
- Rename `OperatorNodeTerminateAndReplace` class -> `DisruptKubernetesNodeThenReplaceScyllaNode`.
- Rename `OperatorNodeTerminateDecommissionAdd` class -> `DisruptKubernetesNodeThenDecommissionAndAddScyllaNode`.

Reason:
Above is done because we were using 3 different disruption methods as one. It is not correct, we should treat them as separate ones.

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: sdcm/nemesis.py
Comment (fruch) [inline sdcm/nemesis.py:1340]: two issues referenced here, which one is the reason we are skipping ?

File: sdcm/nemesis.py
Comment (fruch) [inline sdcm/nemesis.py:1364]: Is this note still relevant? That issue still reproduce?

File: test-cases/scylla-operator/longevity-scylla-operator-3h-terminate-replace.yaml
Comment (fruch) [inline test-cases/scylla-operator/longevity-scylla-operator-3h-terminate-replace.yaml:10]: If we have 3 different nemesis should we have 3 different cases ? Or one nemesis set that would run them all ?

**Review by fruch — COMMENTED** [review]
In general looks o.k.

But I don't want to skip because a new issue opened, I'd rather wait for some feedback on it.

File: sdcm/nemesis.py
Comment (vponomaryov) [reply sdcm/nemesis.py:1340]: The one used in the exception. Second appears also in case of decommission and it doesn't harm there.
So, the duplicated PVs bug is kind of cosmetic for now, as I can see.

File: sdcm/nemesis.py
Comment (vponomaryov) [reply sdcm/nemesis.py:1364]: Whole change is actual. The duplicated PVs bug is still existing, but it is harmless as I can see. THe decommission nemesis has it too, but passes ok.

File: test-cases/scylla-operator/longevity-scylla-operator-3h-terminate-replace.yaml
Comment (vponomaryov) [reply test-cases/scylla-operator/longevity-scylla-operator-3h-terminate-replace.yaml:10]: The very first patch-set to this PR did exactly as you propose in the last statement. 
It was making unit tests fail.

Why?
Because creation of a nemesis with 3 methods causes `SisyphusMonkey` one ignore it's methods for itself when a nemesis filter is provided.

The cause for it is very strange and obfuscated logic for the nemesis methods handling.

Just imagine: SisyphusMonkey class, when is provided with nemesis filter, parses `string` variants of other nemesis classes definitions and looks for the "self.disrupt_foo(" stuff. Find `first` line and returns. It will pass if I have `comment` with such a string in class definition. Then, it returns only first occurrence. All other are ignored.

File: sdcm/nemesis.py
Comment (fruch) [reply sdcm/nemesis.py:1340]: If it's harmless bug, let's drop this comment it doesn't help us.

File: test-cases/scylla-operator/longevity-scylla-operator-3h-terminate-replace.yaml
Comment (fruch) [reply test-cases/scylla-operator/longevity-scylla-operator-3h-terminate-replace.yaml:10]: There a list of classes that should be ignored in that scan.

Comment (fruch) [reply test-cases/scylla-operator/longevity-scylla-operator-3h-terminate-replace.yaml:10]: https://github.com/scylladb/scylla-cluster-tests/blob/1af3cc0f45c0d508b778376c260872fb39f89038/sdcm/nemesis.py#L4876

Comment (vponomaryov) [reply test-cases/scylla-operator/longevity-scylla-operator-3h-terminate-replace.yaml:10]: > There a list of classes that should be ignored in that scan.

If it is ignored that these methods won't be picked up by the `SisyphusMonkey` complex nemesis. But I would expect it to be picked up.
The current logic, without changes, doesn't solve the current need.

Comment (vponomaryov) [reply test-cases/scylla-operator/longevity-scylla-operator-3h-terminate-replace.yaml:10]: Ok, there is a way. Create A simple nemesis class for each of the methods and 2 complex ones. In that case we will be able to get new methods in both cases - only them and as part of the `SisyphusMonkey` one.

But the problem with it is that we add redundant code.

**Comment by vponomaryov** [comment]
> In general looks o.k.
> 
> But I don't want to skip because a new issue opened, I'd rather wait for some feedback on it.

That one skipped method makes cluster lose one node forever. It will break any test run with the `SisyphusMonkey` nemesis.
So, it `must` be skipped.

**Comment by vponomaryov** [comment]
@fruch 

Skipping the disruption method that is skipped here is the current approach in the `master` branch.
This PR is the `dependency` for another change which implements the following task: https://github.com/scylladb/qa-tasks/issues/935

I propose to merge this one if we want to have such separation in general.

**Comment by fruch** [comment]
@vponomaryov please rebase this one

since it introduces new nemesis, now there are tests and nemesis list to update.

**Comment by fruch** [comment]
> @fruch
> 
> Skipping the disruption method that is skipped here is the current approach in the `master` branch. This PR is the `dependency` for another change which implements the following task: [scylladb/qa-tasks#935](https://github.com/scylladb/qa-tasks/issues/935)
> 
> I propose to merge this one if we want to have such disruption methods separation in general.

We do want this, but still I don't like the fact we are skipping some parts, before we have a clue what is the cause and when/if it's going to be fixed

**Comment by vponomaryov** [comment]
> > @fruch
> > Skipping the disruption method that is skipped here is the current approach in the `master` branch. This PR is the `dependency` for another change which implements the following task: [scylladb/qa-tasks#935](https://github.com/scylladb/qa-tasks/issues/935)
> > I propose to merge this one if we want to have such disruption methods separation in general.
> 
> We do want this, but still I don't like the fact we are skipping some parts, before we have a clue what is the cause and when/if it's going to be fixed

Honestly, I don't understand what is that `before` you are referring to here.
This PR doesn't change the number of nemesis we run and/or skip.
It only splits complex nemesis into several simpler ones.

**Comment by fruch** [comment]
> > > @fruch
> > > Skipping the disruption method that is skipped here is the current approach in the `master` branch. This PR is the `dependency` for another change which implements the following task: [scylladb/qa-tasks#935](https://github.com/scylladb/qa-tasks/issues/935)
> > > I propose to merge this one if we want to have such disruption methods separation in general.
> > 
> > We do want this, but still I don't like the fact we are skipping some parts, before we have a clue what is the cause and when/if it's going to be fixed
> 
> Honestly, I don't understand what is that `before` you are referring to here.
> This PR doesn't change the number of nemesis we run and/or skip.
> It only splits complex nemesis into several simpler ones.

Please rebase.

The before, is that I don't like the fact we are skipping important case while we have zero clue of the cause.

Last time that it was skipped it was forgotten about for almost a whole year, we need to push it forward, our work doesn't end in just opening an issue.

**Comment by vponomaryov** [comment]
> > > > @fruch
> > > > Skipping the disruption method that is skipped here is the current approach in the `master` branch. This PR is the `dependency` for another change which implements the following task: [scylladb/qa-tasks#935](https://github.com/scylladb/qa-tasks/issues/935)
> > > > I propose to merge this one if we want to have such disruption methods separation in general.
> > > 
> > > 
> > > We do want this, but still I don't like the fact we are skipping some parts, before we have a clue what is the cause and when/if it's going to be fixed
> > 
> > 
> > Honestly, I don't understand what is that `before` you are referring to here.
> > This PR doesn't change the number of nemesis we run and/or skip.
> > It only splits complex nemesis into several simpler ones.
> 
> Please rebase.
> 
> The before, is that I don't like the fact we are skipping important case while we have zero clue of the cause.
> 
> Last time that it was skipped it was forgotten about for almost a whole year, we need to push it forward, our work doesn't end in just opening an issue.

Forgotten by whom? The case is still broken.

At first, what do you propose to do?
At second, the discussion about the `skip/unskip` is unrelated to this particular change.

**Comment by soyacz** [comment]
Maybe when doing this split, we could unify some naming, e.g. put `k8s` or `kubernetes` after word `disrupt_`? e.g: `disrupt_kubernetes_replace_scylla_node`

**Comment by fruch** [comment]
> Maybe when doing this split, we could unify some naming, e.g. put `k8s` or `kubernetes` after word `disrupt_`? e.g: `disrupt_kubernetes_replace_scylla_node`

those are gonna by label as such in following work, why do we need to prefix the names ?

**Comment by fruch** [comment]
> > > > > @fruch
> > > > > Skipping the disruption method that is skipped here is the current approach in the `master` branch. This PR is the `dependency` for another change which implements the following task: [scylladb/qa-tasks#935](https://github.com/scylladb/qa-tasks/issues/935)
> > > > > I propose to merge this one if we want to have such disruption methods separation in general.
> > > > 
> > > > 
> > > > We do want this, but still I don't like the fact we are skipping some parts, before we have a clue what is the cause and when/if it's going to be fixed
> > > 
> > > 
> > > Honestly, I don't understand what is that `before` you are referring to here.
> > > This PR doesn't change the number of nemesis we run and/or skip.
> > > It only splits complex nemesis into several simpler ones.
> > 
> > 
> > Please rebase.
> > The before, is that I don't like the fact we are skipping important case while we have zero clue of the cause.
> > Last time that it was skipped it was forgotten about for almost a whole year, we need to push it forward, our work doesn't end in just opening an issue.
> 
> Forgotten by whom? The case is still broken.

Forgotten by us, in
https://github.com/vponomaryov/scylla-cluster-tests/commit/678eeb6c973e9c0d18f2ff42b7a48aaaea2f93c6
we replaced the issues, and kept on 

now we facing a new issue, and we add it as skipped in a different manner.

> 
> At first, what do you propose to do? At second, the discussion about the `skip/unskip` is unrelated to this particular change.

it's related  to the PR, since we are introducing a new skip here.

Let's see what we can do to push forward resolving that issue issue, don't want to wait a year an a half, assuming it would fix itself.

File: sdcm/nemesis.py
Comment (fruch) [inline sdcm/nemesis.py:1340]: now that we know it's working as we expect in serverless, let's not skip it for serverless

**Review by fruch — CHANGES_REQUESTED** [review]
- remove the skip from disrupt_terminate_kubernetes_host_then_replace_scylla_node for serverless
- rebase this PR

**Comment by soyacz** [comment]
> > Maybe when doing this split, we could unify some naming, e.g. put `k8s` or `kubernetes` after word `disrupt_`? e.g: `disrupt_kubernetes_replace_scylla_node`
> 
> those are gonna by label as such in following work, why do we need to prefix the names ?

I was just thinking, when we already change names, so maybe put some order to naming. It's just suggestion.

**Comment by vponomaryov** [comment]
> > > Maybe when doing this split, we could unify some naming, e.g. put `k8s` or `kubernetes` after word `disrupt_`? e.g: `disrupt_kubernetes_replace_scylla_node`
> > 
> > 
> > those are gonna by label as such in following work, why do we need to prefix the names ?
> 
> I was just thinking, when we already change names, so maybe put some order to naming. It's just suggestion.

The current approach has logic. First part is required `disrupt_` prefix, then goes `action type` and then the `object` for action.

**Review by fruch — APPROVED** [review]
LGTM

**Comment by vponomaryov** [comment]
@fruch it requires your attention

**Review by fruch — APPROVED** [review]
LGTM

**Pull request merged**
